src/protocol/attribute.rs

```diff
@@ -0,0 +1,224 @@
+            let mut b = self
+                .multiply_all(
+                    &input[Self::_HELPER_BIT][step_size..],
+                    &input[Self::_IS_TRIGGER_BIT][step_size..],
+                )
+                .await?;
+
+            // since `stop_bits` is initialized with `1`s, we only multiply [stop_bit] on the second and later iterations
+            if step_size > 1 {
+                b = self.multiply_all(&stop_bits[..end], &b).await?;
+            }
+
+            // [b * successor.credit]
+            let b_cred = self.multiply_all(&b[..end], &credits[step_size..]).await?;
+
+            // [credit] = [credit + b * successor.credit]
+            b_cred
+                .into_iter()
+                .enumerate()
+                .for_each(|(i, x)| credits[i] += x);
+
+            // [stop_bit] = [b * successor.stop_bit]
+            // for the same reason as calculating [b], we skip the multiplication on the first iteration
+            let b_stop = if step_size > 1 {
+                self.multiply_all(&b[..end], &stop_bits[step_size..])
+                    .await?
+            } else {
+                b.clone()
+            };
+
+            b_stop
+                .into_iter()
+                .enumerate()
+                .for_each(|(i, x)| stop_bits[i] = x);
+
+            step_size *= 2;
+        }
+
+        Ok(credits)
+    }
+
+    async fn multiply_all<M, F>(
+        &self,
+        a: &[Replicated<F>],
+        b: &[Replicated<F>],
+    ) -> Result<Vec<Replicated<F>>, BoxError>
+    where
+        F: Field,
+        M: Mesh,
+        G: Gateway<M, AttributionModel>,
+    {
+        assert_eq!(a.len(), b.len());
+
+        try_join_all((0..a.len()).map(|i| async move {
+            let x = a[i];
+            let y = b[i];
+
+            // assume # of records < 2^32
+            #[allow(clippy::cast_possible_truncation)]
+            self.ctx
+                .multiply(RecordId::from(i as u32), self.step)
+                .await
+                .execute(x, y)
+                .await
+        }))
+        .await
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::{
+        field::{Field, Fp31},
+        protocol::{attribute::Attribute, AttributionModel, QueryId},
+        test_fixture::{make_contexts, make_world, share, validate_and_reconstruct, TestWorld},
+    };
+    use rand::rngs::mock::StepRng;
+    use tokio::try_join;
+
+    #[tokio::test]
+    pub async fn accumulate() {
+        let world: TestWorld<AttributionModel> = make_world(QueryId);
+        let context = make_contexts(&world);
+        let mut rng = StepRng::new(100, 1);
+
+        let input: [[u128; 9]; 6] = [
+            [1, 1, 1, 1, 1, 1, 1, 2, 3],
+            [2, 5, 8, 9, 10, 13, 27, 11, 14],
+            [0, 0, 1, 1, 1, 1, 1, 0, 1],
+            [0, 1, 1, 1, 1, 1, 1, 0, 0],
+            [3, 4, 0, 0, 0, 0, 0, 1, 0],
+            [0, 0, 10, 2, 1, 5, 1, 0, 10],
+        ];
```

Comment:
Yeah, I have no idea what this means. I think the structs will help.

Reply:
Yeah, this is my fault. I didn't add any comments to this and some arrays are irrelevant. I'll clarify in this thread. https://github.com/martinthomson/raw-ipa/pull/108#pullrequestreview-1136724098